### PR TITLE
mremoteng: Use portable configuration

### PR DESCRIPTION
**Pull request by kraptor**
mRemoteNG default installer fails on very limited user accounts where organizations prevent access to .msi files/installers (via antivirus software or by other means).

In order to allow mRemoteNG to install, one can install the "portable" version (which is a simple ZIP file) and persist some of its configuration.

NOTE: mremoteng-portable.json is based on mremoteng.json.

### CODE REVIEW

**Comment by kraptor** [comment]
I upsated the commit to pass all style checks from AppVeyor. Let me know if there is anything else to do.

File: bucket/mremoteng-portable.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    ],
+    "persist": [
+        "mRemoteNG.settings",
+        "mRemoteNG.log",
+        "extApps.xml",
+        "pnlLayout.xml"
+    ],
+    "notes": [
+        "Default connections file is not persisted! Use a custom connection file insted.",
+        "On first run, you can load the default panel layout by using 'View'->'Reset Layout'."
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]: Remove your layout. Nonse to keep user preference on single user

File: bucket/mremoteng-portable.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
...e='DockLeft' ActiveContent='1'>      <Contents Count='1'>        <Content ID='0' RefID='1' />      </Contents>    </Pane>    <Pane ID='2' DockState='DockBottomAutoHide' ActiveContent='2'>      <Contents Count='1'>        <Content ID='0' RefID='2' />      </Contents>    </Pane>  </Panes>  <DockWindows>    <DockWindow ID='0' DockState='Document' ZOrderIndex='1'>      <NestedPanes Count='0' />    </DockWindow>    <DockWindow ID='1' DockState='DockLeft' ZOrderIndex='2'>      <NestedPanes Count='2'>        <Pane ID='0' RefID='0' PrevPane='-1' Alignment='Bottom' Proportion='0.5' />        <Pane ID='1' RefID='1' PrevPane='0' Alignment='Bottom' Proportion='0.5' />      </NestedPanes>    </DockWindow>    <DockWindow ID='2' DockState='DockRight' ZOrderIndex='3'>      <NestedPanes Count='0' />    </DockWindow>    <DockWindow ID='3' DockState='DockTop' ZOrderIndex='4'>      <NestedPanes Count='0' />    </DockWindow>    <DockWindow ID='4' DockState='DockBottom' ZOrderIndex='5'>      <NestedPanes Count='1'>        <Pane ID='0' RefID='2' PrevPane='-1' Alignment='Right' Proportion='0.5' />      </NestedPanes>    </DockWindow>  </DockWindows>  <FloatWindows Count='0' /></DockPanel>\"  | Out-Null }"
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]: Create just empty file and remove that nonsense layout.

File: bucket/mremoteng-portable.json [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    "shortcuts": [
+        [
+            "mRemoteNG.exe",
+            "mRemoteNG"
+        ]
+    ],
+    "pre_install": [
+        "if (!(Test-Path \"$persist_dir\\mRemoteNG.settings\")) { New-Item \"$dir\\mRemoteNG.settings\" -ItemType File -Value \"<?xml version='1.0' encoding='utf-8'?>\"  | Out-Null }",
+        "if (!(Test-Path \"$persist_dir\\mRemoteNG.log\"  )) { New-Item \"$dir\\mRemoteNG.log\" -ItemType File | Out-Null }",
+        "if (!(Test-Path \"$persist_dir\\extApps.xml\"    )) { New-Item \"$dir\\extApps.xml\"   -ItemType File -Value \"<?xml version='1.0' encoding='utf-8'?>\"  | Out-Null }",
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]: What is this for?

File: bucket/mremoteng-portable.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    "notes": [
+        "Default connections file is not persisted! Use a custom connection file insted.",
+        "On first run, you can load the default panel layout by using 'View'->'Reset Layout'."
+    ],
+    "checkver": {
+        "github": "https://github.com/mRemoteNG/mRemoteNG",
+        "re": "/releases/download/(?<short>[^/]+)/mRemoteNG-Portable-(?<version>[\\d.]+).zip"
+    },
+    "autoupdate": {
+        "url": "https://github.com/mRemoteNG/mRemoteNG/releases/download/$matchShort/mRemoteNG-Porable-$version.zip"
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]:
```suggestion
        "url": "https://github.com/mRemoteNG/mRemoteNG/releases/download/v$matchHead/mRemoteNG-Portable-$version.zip"
```

File: bucket/mremoteng-portable.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        "extApps.xml",
+        "pnlLayout.xml"
+    ],
+    "notes": [
+        "Default connections file is not persisted! Use a custom connection file insted.",
+        "On first run, you can load the default panel layout by using 'View'->'Reset Layout'."
+    ],
+    "checkver": {
+        "github": "https://github.com/mRemoteNG/mRemoteNG",
+        "re": "/releases/download/(?<short>[^/]+)/mRemoteNG-Portable-(?<version>[\\d.]+).zip"
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]:
```suggestion
        "regex": "mRemoteNG-Portable-([\\d.]+)\\.zip"
```

File: bucket/mremoteng-portable.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
+{
+    "homepage": "https://mremoteng.org/",
+    "description": "An open source, tabbed, multi-protocol, remote connections manager.",
+    "version": "1.76.20.24669",
+    "hash": "97b0e94bbdc409f3eb09b1017b3b58ccc0eae145a98a3bc6ea651ef1d5bc2d8e",
+    "license": "GPL-2.0-only",
+    "url": "https://github.com/mRemoteNG/mRemoteNG/releases/download/v1.76.20/mRemoteNG-Portable-1.76.20.24669.zip",
+    "bin": "mRemoteNG.exe",
+    "extract_dir": "",
```
Comment (Ash258) [inline bucket/mremoteng-portable.json]:
```suggestion
```

File: bucket/mremoteng-portable.json (hunk 2 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: sadly, mRemoteNG fails (errors out with a message box) on empty files :(

File: bucket/mremoteng-portable.json (hunk 1 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: This is the default layout. Empty layout means mRemoteNG appears without any panels. I forgot to remove the second message. If we keep the message we can just have an empty layout.

File: bucket/mremoteng-portable.json (hunk 3 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: Default empty files to keep mRemoteNG happy and not failing (or loading empty data) which is not the default on a new installation. We can review thoroughly if needed.

File: bucket/mremoteng-portable.json (hunk 4 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: I agree, I'll review once I get home.

File: bucket/mremoteng-portable.json (hunk 5 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: I copied what the original one had... I agree this looks better.

File: bucket/mremoteng-portable.json (hunk 6 above)
Comment (kraptor) [reply bucket/mremoteng-portable.json]: For some reason it failed when not set... may be I did something wrong. When I added the key with "", then it worked ok.

**Comment by Ash258** [comment]
- Fixed manifest
- Backward compatible with automatic config moving from appdata
- Copy over method instead of creating errorprone configurations

<details>
<summary>Manifest</summary>

```json
{
    "version": "1.76.20.24669",
    "description": "Fork of mRemote; Tabbed, multi-protocol remote connections manager",
    "homepage": "https://mremoteng.org",
    "license": "GPL-2.0-only",
    "url": "https://github.com/mRemoteNG/mRemoteNG/releases/download/v1.76.20/mRemoteNG-Portable-1.76.20.24669.zip",
    "hash": "97b0e94bbdc409f3eb09b1017b3b58ccc0eae145a98a3bc6ea651ef1d5bc2d8e",
    "pre_install": [
        "$loc = \"$env:LOCALAPPDATA\\mRemoteNG\"",
        "$roam = \"$env:APPDATA\\mRemoteNG\"",
        "    # Copy user configuration",
        "if (Test-Path $loc) {",
        "    $file = (Get-ChildItem \"$loc\\*\" 'user.config' -File -Recurse | Sort-Object -Property LastWriteTime)[-1].FullName",
        "    Copy-Item $file \"$persist_dir\\mRemoteNG.settings\"",
        "    Move-Item $loc \"$loc.Backup\"",
        "}",
        "    # Migrate user preferences",
        "if (Test-Path $roam) {",
        "    Get-ChildItem \"$roam\\*\" -File -Exclude '*.backup' | Copy-Item -Destination \"$persist_dir\"",
        "    Move-Item $roam \"$roam.Backup\"",
        "}",
        "",
        "if (-not (Test-Path \"$persist_dir\\mRemoteNG.settings\")) {",
        "    Set-Content -Path \"$dir\\mRemoteNG.settings\" -Encoding Ascii -Value @'",
        "<?xml version=\"1.0\" encoding=\"utf-8\"?>",
        "<settings>",
        "  <localSettings>",
        "     <setting name=\"CheckForUpdatesOnStartup\">False</setting>",
        "     <setting name=\"CheckForUpdatesAsked\">True</setting>",
        "  </localSettings>",
        "</settings>",
        "'@",
        "}",
        "",
        "'pnlLayout.xml', 'extApps', 'confCons.xml' | ForEach-Object {",
        "    Copy-Item \"$persist_dir\\$_\" \"$dir\" -ErrorAction 'SilentlyContinue' -Force",
        "}"
    ],
    "uninstaller": {
        "script": [
            "'pnlLayout.xml', 'extApps', 'confCons.xml' | ForEach-Object {",
            "    Copy-Item \"$dir\\$_\" \"$persist_dir\" -ErrorAction 'SilentlyContinue' -Force",
            "}"
        ]
    },
    "bin": [
        "mRemoteNG.exe",
        "PuTTYNG.exe"
    ],
    "shortcuts": [
        [
            "mRemoteNG.exe",
            "mRemoteNG"
        ],
        [
            "PuTTYNG.exe",
            "PuTTYNG"
        ]
    ],
    "persist": "mRemoteNG.settings",
    "checkver": {
        "github": "https://github.com/mRemoteNG/mRemoteNG",
        "regex": "mRemoteNG-Portable-([\\d.]+)\\.zip"
    },
    "autoupdate": {
        "url": "https://github.com/mRemoteNG/mRemoteNG/releases/download/v$matchHead/mRemoteNG-Portable-$version.zip"
    }
}

```

</details>

**Comment by kraptor** [comment]
@Ash258 how do we proceed? Can you create another pull request with your manifeest so I can take this one down?

**Comment by Ash258** [comment]
No. Just copy the json in comment into file and push

**Comment by kraptor** [comment]
I did, but not fair to merge as @Ash258 loses authorship... 

**Please, some maintainer review this before merging.**

**Comment by kraptor** [comment]
Fixed issues found by CI.

**Comment by kraptor** [comment]
Any update on this?

**Pull request closed without merging**